### PR TITLE
refactor: move from Converter to GenConverter

### PR DESCRIPTION
**Pull request by aarnphm**
This should be compatible with current GenConverter.

address #2684

### CODE REVIEW

File: bentoml/_internal/bento/bento.py [hunk 1]
```diff
@@ -340,74 +341,62 @@ def __init__(self, base_path: t.Union[PathType, "FS"]):
...
     name: str
     runnable_type: str
     models: t.List[str] = attr.field(factory=list)
     resource_config: t.Optional[t.Dict[str, t.Any]] = attr.field(default=None)
 
     @classmethod
-    def from_runner(cls, r: Runner) -> "BentoRunnerInfo":
-        # Add runner default resource quota and batching config here
+    def from_runner(cls, r: Runner) -> BentoRunnerInfo:
+        # TODO: Add runner default resource quota and batching config here
```
Comment (parano) [inline bentoml/_internal/bento/bento.py:353]:
```suggestion
```

File: requirements/dev-requirements.txt [hunk 2]
```diff
@@ -1,5 +1,6 @@
 # Dev dependencies
 -r tests-requirements.txt
+-r docs-requirements.txt
```
Comment (parano) [inline requirements/dev-requirements.txt:3]:
```suggestion
```

File: bentoml/_internal/utils/cattr.py [hunk 3]
```diff
@@ -3,12 +3,18 @@
...
-import cattr
 from attr import fields
-from cattr import override  # type: ignore
+from cattr.gen import override
 from cattr.gen import AttributeOverride
 
-bentoml_cattr = cattr.Converter()
+from .pkg import pkg_version_info
+
+if pkg_version_info("cattrs")[:2] >= (22, 2):
```
Comment (parano) [inline bentoml/_internal/utils/cattr.py:12]: use packaging.version's comparison implementation

File: bentoml/_internal/utils/cattr.py [hunk 4]
```diff
@@ -3,12 +3,18 @@
...
+    from cattr import Converter
+else:
+    from cattr import GenConverter as Converter
```
Comment (parano) [inline bentoml/_internal/utils/cattr.py:15]: This doesn't work at all for version < 22.2

File: setup.cfg [hunk 5]
```diff
@@ -72,6 +72,7 @@ install_requires =
     starlette
     typing-extensions>=4.0
     backports.cached-property;python_version<"3.8"
+    exceptiongroup;python_version <= "3.11"
```
Comment (aarnphm) [inline setup.cfg:74]: This will be the default package for python 3.11. This would actually provide better exception traceback than rich traceback.
Screenshot 2022-07-05 at 00 35 59

Comment (aarnphm) [reply setup.cfg:74]: rich traceback haven't supported this yet. We could allow user to enable rich.traceback if then want to. But I think for now exceptiongroup is more verbose.

File: bentoml/_internal/types.py [hunk 6]
```diff
@@ -48,27 +48,26 @@
...
+ValueType = t.Union[
+    str, bytes, bool, int, float, complex, datetime, date, time, timedelta
+]
+MetadataType = t.Union[
+    ValueType,
+    t.List[ValueType],
+    t.Dict[str, ValueType],
+    t.Dict[t.Union[str, float, int], ValueType],
+    t.Tuple[ValueType],
 ]
```
Comment (parano) [inline bentoml/_internal/types.py:64]: Is this change necessary? it's not exactly correct.

 I think only the l66-70 is what's required for cattrs to work.

File: bentoml/_internal/cli/__init__.py [hunk 7]
```diff
@@ -12,11 +12,6 @@
 
 
 def create_bentoml_cli():
-    # exclude traceback from the click library
-    from rich.traceback import install
-
-    install(suppress=[click])
-
```
Comment (aarnphm) [inline bentoml/_internal/cli/__init__.py:19]: using exceptiongroup returns better exception message

File: setup.cfg (hunk 5 above)
Comment (aarnphm) [reply setup.cfg:74]: We should move back to rich.traceback once https://github.com/Textualize/rich/issues/1859 supports it :)

File: bentoml/_internal/types.py (hunk 6 above)
Comment (aarnphm) [reply bentoml/_internal/types.py:64]: oh no, let me revert this back. 

You are correct. Only L66-70 is needed.

File: bentoml/_internal/utils/cattr.py [hunk 8]
```diff
@@ -3,12 +3,18 @@
...
 
-bentoml_cattr = cattr.Converter()
+from .pkg import pkg_version_info
+
+if pkg_version_info("cattrs") >= (22, 2):
+    from cattr import Converter
+else:
+    from cattr import GenConverter as Converter
+
+bentoml_cattr = Converter(forbid_extra_keys=True)
```
[the remark this answers is not in the document]
Comment (parano) [reply bentoml/_internal/utils/cattr.py:17]: that was the whole point ;)

[the remark this answers is not in the document]
Comment (aarnphm) [reply bentoml/_internal/utils/cattr.py:17]: should work as expected. on my end it works with both python 3.7 and python 3.10

[the remark this answers is not in the document]
Comment (parano) [reply bentoml/_internal/utils/cattr.py:17]: > wait wdym? we want to forbid_extra_keys. This works with 22.1.0 =='

forbid_extra_keys=True is the default behavior in BaseConvertor, we need forbid_extra_keys=False for validating user provided yaml config

[the remark this answers is not in the document]
Comment (ssheng) [reply bentoml/_internal/utils/cattr.py:17]: If we want different `forbid_extra_keys` behaviors based on the `cattrs` version, should we put the converter initialization inside the if statement as well?

**Comment by ssheng** [comment]
@aarnphm do you think this is needed for the 1.0 release? On a separate topic, to help focus the PR, I'd recommend that we separate the type refactoring and the functional change. 😃

**Comment by aarnphm** [comment]
> @aarnphm do you think this is needed for the 1.0 release? On a separate topic, to help focus on the actual issue, I'd recommend that we separate the type refactoring and the functional change. 😃

This is needed for 1.0

File: setup.cfg [hunk 9]
```diff
@@ -39,6 +39,7 @@ install_requires =
     aiohttp<=3.8.1  # prevent it from installing preview release
     attrs>=21.1.0
     cattrs>=1.10.0,<=21.1.0
+>>>>>>> 0ffe72d5 (chore: updates types for lower python version)
```
Comment (aarnphm) [inline setup.cfg:42]:
```suggestion
```

File: bentoml/_internal/bento/bento.py [hunk 10]
```diff
@@ -340,81 +341,69 @@ def __init__(self, base_path: t.Union[PathType, "FS"]):
         super().__init__(base_path, Bento)
 
 
-@attr.define(frozen=True, on_setattr=None)
```
Comment (sauyon) [inline bentoml/_internal/bento/bento.py:343]: Why did you remove this? This was for compatibility with some (probably older) `attrs` version, I think.

We should definitely add a comment as explanation here, though.

Comment (aarnphm) [reply bentoml/_internal/bento/bento.py:343]: define is a new 22.1 feature anyway right?

Comment (sauyon) [reply bentoml/_internal/bento/bento.py:343]: Sure, but there was at least one version where not doing this did something we didn't want it to do. Don't remember which one, though...

Comment (sauyon) [reply bentoml/_internal/bento/bento.py:343]: Seems like it's in from v 21.

Comment (aarnphm) [reply bentoml/_internal/bento/bento.py:343]: I mean don't we already do `@attr.frozen` anyway?

File: bentoml/_internal/bento/bento.py [hunk 11]
```diff
@@ -340,75 +341,69 @@ def __init__(self, base_path: t.Union[PathType, "FS"]):
...
     service: str = attr.field(
         converter=lambda svc: svc if isinstance(svc, str) else svc._import_str
-    )  # type: ignore[reportPrivateUsage]
-    name: str = attr.field(init=False)  # converted from tag in __attrs_post_init__
-    version: str = attr.field(init=False)  # converted from tag in __attrs_post_init__
-    bentoml_version: str = attr.field(default=BENTOML_VERSION)
+    )
+    name: str = attr.field(init=False)
+    version: str = attr.field(init=False)
+    # NOTE: using factory explicitly instead of default because omit_if_default is enabled globally in bentoml_cattr
```
Comment (aarnphm) [inline bentoml/_internal/bento/bento.py:405]:
```suggestion
    # using factory explicitly instead of default because omit_if_default is enabled for BentoInfo
```

File: bentoml/_internal/models/model.py [hunk 12]
```diff
@@ -318,8 +318,10 @@ def __init__(self, base_path: "t.Union[PathType, FS]"):
 class ModelContext:
     framework_name: str
     framework_versions: t.Dict[str, str]
-    bentoml_version: str = attr.field(default=BENTOML_VERSION)
-    python_version: str = attr.field(default=PYTHON_VERSION)
+
+    # NOTE: using factory explicitly instead of default because omit_if_default is enabled globally in bentoml_cattr
```
Comment (aarnphm) [inline bentoml/_internal/models/model.py:322]:
```suggestion
    # using factory explicitly instead of default because omit_if_default is enabled in ModelInfo

File: bentoml/_internal/bento/bento.py (hunk 10 above)
Comment (parano) [reply bentoml/_internal/bento/bento.py:343]: @sauyon we have `attrs>=21.1.0` in our dependency now, should be ok to just use `@attr.frozen` here now?

File: bentoml/_internal/bento/build_config.py [hunk 13]
```diff
@@ -132,6 +129,12 @@ def _convert_env(
 
 @attr.frozen
 class DockerOptions:
+
+    # User shouldn't add new fields under yaml file.
```
Comment (parano) [inline bentoml/_internal/bento/build_config.py:133]:
```suggestion
    # For validating user defined bentofile.yaml.
```
```suggestion
    # User shouldn't add new fields under yaml file.
```
```suggestion
    # User shouldn't add new fields under yaml file.
```

File: bentoml/_internal/bento/build_config.py [hunk 14]
```diff
@@ -132,6 +129,12 @@ def _convert_env(
...
+    __forbid_extra_keys__ = True
+    # no need to omit since BentoML has already handled the default values.
```
Comment (parano) [inline bentoml/_internal/bento/build_config.py:135]:
```suggestion
    # always omit config values in case of default values got changed in future BentoML releases
```
```suggestion
    # no need to omit since BentoML has already handled the default values.
```
```suggestion
    # no need to omit since BentoML has already handled the default values.
```

File: bentoml/_internal/utils/cattr.py [hunk 15]
```diff
@@ -28,5 +26,23 @@ def datetime_decoder(dt_like: str | datetime, _: t.Any) -> datetime:
...
 
 
-bentoml_cattr.register_unstructure_hook(datetime, lambda dt: dt.isoformat())  # type: ignore
-bentoml_cattr.register_structure_hook(datetime, datetime_decoder)
+bentoml_cattr.register_structure_hook_factory(
+    attr.has,
+    lambda cls: make_dict_structure_fn(
+        cls,
+        bentoml_cattr,
+        _cattrs_forbid_extra_keys=getattr(cls, "__forbid_extra_keys__", False),
```
Comment (ssheng) [inline bentoml/_internal/utils/cattr.py:34]: I read that dunders (double leading and trailing underscore) should be reserved for python internals. Unlikely we will get a collision with this name but should we by consistent with naming and maybe use ALL_UPPERCASE or _single_leading_underscore?

**Review by ssheng — APPROVED** [review]
One minor comment.

File: bentoml/_internal/utils/cattr.py (hunk 15 above)
Comment (aarnphm) [reply bentoml/_internal/utils/cattr.py:34]: I think dunder usage in this case is fine, since this is only specific to a bentoml internal. The best way is to use metaclass, but then there is no need for that complexity.
